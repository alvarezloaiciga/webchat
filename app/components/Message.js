// @flow

import React from 'react';
import QUIQ, {getStyle} from 'utils/quiq';
import classnames from 'classnames';
import Linkify from 'react-linkify';
import type {Message as MessageType} from 'types';
import './styles/Message.scss';

export type MessageProps = {
  message: MessageType,
};

const {FONT_FAMILY, COLORS, STYLES, WIDTH} = QUIQ;

const getMessageBubbleStyle = (fromCustomer: boolean) => {
  if (fromCustomer) {
    return getStyle(STYLES.CustomerMessageBubble, {
      backgroundColor: COLORS.customerMessageBackground,
      maxWidth: WIDTH - 50,
    });
  }

  return getStyle(STYLES.AgentMessageBubble, {
    backgroundColor: COLORS.agentMessageBackground,
    maxWidth: WIDTH - 50,
  });
};

const getMessageTextStyle = (fromCustomer: boolean) => {
  if (fromCustomer) {
    return getStyle(STYLES.CustomerMessageText, {
      color: COLORS.customerMessageText,
      fontFamily: FONT_FAMILY,
    });
  }

  return getStyle(STYLES.AgentMessageText, {
    color: COLORS.agentMessageText,
    fontFamily: FONT_FAMILY,
  });
};

export const Message = (props: MessageProps) => {
  const fromCustomer = props.message.authorType === 'Customer';

  const textStyle = getMessageTextStyle(fromCustomer);

  const margin = fromCustomer
    ? {marginLeft: 'auto', justifyContent: 'flex-end'}
    : {marginRight: 'auto', justifyContent: 'flex-start'};

  return (
    <div className={classnames('messageContainer', {fromCustomer})}>
      <div style={margin}>
        {!fromCustomer && <div className="agentAvatar" style={getStyle(STYLES.AgentAvatar)} />}
        <div
          style={getMessageBubbleStyle(fromCustomer)}
          className={classnames('Message', {fromCustomer})}
        >
          <Linkify
            properties={{
              target: '_blank',
              rel: 'noopener noreferrer',
              style: {
                fontFamily: FONT_FAMILY,
                textDecoration: 'underline',
                ...textStyle,
                color: fromCustomer ? COLORS.customerMessageLinkText : COLORS.agentMessageLinkText,
              },
            }}
          >
            <span style={textStyle}>
              {props.message.text}
            </span>
          </Linkify>
        </div>
        {fromCustomer && <div className="customerAvatar" style={getStyle(STYLES.CustomerAvatar)} />}
      </div>
    </div>
  );
};

export default Message;
