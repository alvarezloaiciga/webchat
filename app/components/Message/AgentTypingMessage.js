// @flow
import React from 'react';
import quiqOptions, {getStyle, getMessage} from 'Common/QuiqOptions';
import {messageTypes} from 'Common/Constants';
import TypingIndicator from 'TypingIndicator';

export type AgentTypingMessageProps = {
  scrollToBottom: () => void,
};

export class AgentTypingMessage extends React.Component<AgentTypingMessageProps> {
  props: AgentTypingMessageProps;

  componentDidMount() {
    setTimeout(this.props.scrollToBottom, 200);
  }

  render() {
    const {colors, styles} = quiqOptions;
    return (
      <TypingIndicator
        title={getMessage(messageTypes.agentTypingMessage)}
        gradientColor={{
          foreground: colors.typingIndicatorForeground,
          background: colors.typingIndicatorBackground,
        }}
        svgStyle={getStyle(styles.TypingIndicatorSvgStyle)}
        circleStyle={getStyle(styles.TypingIndicatorCircleStyle)}
      />
    );
  }
}

export default AgentTypingMessage;
