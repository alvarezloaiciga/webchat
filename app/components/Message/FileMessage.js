// @flow

import React from 'react';
import quiqOptions, {getStyle} from 'Common/QuiqOptions';
import {getMetadataForMimeType} from 'utils/fileUtils';
import classnames from 'classnames';
import ProgressCircle from '../ProgressCircle';
import Icon from 'core-ui/components/Icon';
import type {AttachmentMessage as AttachmentMessageType} from 'Common/types';
import {css} from 'react-emotion';
import {messageEnter} from 'animations';

export type FileMessageProps = {
  message: AttachmentMessageType,
};

const {fontFamily, colors, styles, width} = quiqOptions;

const getMessageBubbleStyle = (fromCustomer: boolean) => {
  if (fromCustomer) {
    return getStyle(styles.CustomerAttachmentBubble, {
      borderColor: colors.attachmentMessageColor,
      maxWidth: width - 82,
    });
  }

  return getStyle(styles.AgentAttachmentBubble, {
    borderColor: colors.attachmentMessageColor,
    maxWidth: width - 50,
  });
};

const getMessageTextStyle = (fromCustomer: boolean) => {
  if (fromCustomer) {
    return getStyle(styles.CustomerAttachmentText, {
      color: colors.attachmentMessageColor,
      fontFamily,
    });
  }

  return getStyle(styles.AgentAttachmentText, {
    color: colors.attachmentMessageColor,
    fontFamily,
  });
};

const FileMessageStyle = css`
  a {
    text-decoration: none;
    color: inherit;
  }

  .FileMessage {
    width: auto;
    flex: 0 1 auto;
    word-wrap: break-word;
    background: transparent;
    border: 1px solid #9c9c9f;
    color: #9c9c9f;
    padding: 8px;
    border-radius: 5px;
    font-size: 14px;
    text-align: left;
    animation: 0.2s 1 ${messageEnter};
    position: relative;
    overflow: hidden;
    hyphens: auto;
    box-sizing: border-box;

    span {
      display: flex;
      align-items: center;
    }

    .IconContainer {
      height: 30px;
      margin-right: 10px;
      display: flex;
      align-items: center;

      .CircularProgressbar {
        display: block;
      }
    }
  }
`;

export const FileMessage = (props: FileMessageProps) => {
  const fromCustomer = props.message.authorType === 'Customer';
  const isUploading = props.message.status && props.message.status === 'pending';
  const textStyle = getMessageTextStyle(fromCustomer);
  const {name, icon} = getMetadataForMimeType(props.message.contentType);

  return (
    <div className={`FileMessageContainer ${FileMessageStyle}`}>
      <a href={props.message.url} target="_blank" rel="noopener noreferrer">
        <div
          style={getMessageBubbleStyle(fromCustomer)}
          className={classnames('FileMessage', {fromCustomer})}
        >
          <span style={textStyle}>
            <div className="IconContainer">
              {isUploading ? (
                <ProgressCircle
                  percentage={props.message.uploadProgress || 0}
                  size={30}
                  progressWidth={2}
                  trackWidth={2}
                  progressColor="#7b7b7d"
                  trackColor="rgba(242, 242, 242, 0.58)"
                />
              ) : (
                <Icon icon={icon} title={name} size="2x" />
              )}
            </div>
            {name}
          </span>
        </div>
      </a>
    </div>
  );
};

export default FileMessage;
