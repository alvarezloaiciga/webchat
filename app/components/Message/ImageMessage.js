// @flow

import React from 'react';
import classnames from 'classnames';
import ProgressCircle from '../ProgressCircle';
import type {AttachmentMessage as AttachmentMessageType} from 'Common/types';
import './styles/ImageMessage.scss';

export type ImageMessageProps = {
  message: AttachmentMessageType,
  scrollToBottom: () => void,
};

export type ImageMessageState = {
  imageLoaded: boolean,
  imageWidth: number,
  imageHeight: number,
};

export class ImageMessage extends React.Component<ImageMessageProps, ImageMessageState> {
  props: ImageMessageProps;
  state: ImageMessageState = {
    imageLoaded: false,
    imageHeight: 0,
    imageWidth: 0,
  };
  image: ?Image;
  pollingInterval: IntervalID;

  componentWillMount() {
    this.loadImage(this.props.message.localBlobUrl || this.props.message.url);
  }

  componentWillReceiveProps(nextProps: ImageMessageProps) {
    // Start loading new image if one of the sources changed
    if (
      nextProps.message.localBlobUrl &&
      nextProps.message.localBlobUrl !== this.props.message.localBlobUrl
    ) {
      this.loadImage(nextProps.message.localBlobUrl);
    } else if (
      !nextProps.message.localBlobUrl &&
      nextProps.message.url &&
      nextProps.message.url !== this.props.message.url
    ) {
      this.loadImage(nextProps.message.url);
    }
  }

  componentDidUpdate(prevProps: ImageMessageProps, prevState: ImageMessageState) {
    if (this.state.imageHeight && !prevState.imageHeight) {
      this.props.scrollToBottom();
    }
  }

  loadImage = (url: string) => {
    this.setState({imageLoaded: false});
    this.image = new Image();
    this.image.src = url;
    if (this.image.complete) {
      this.handleImageLoad();
    } else {
      // Listen for image load
      this.image.addEventListener('load', this.handleImageLoad);

      // Poll for dimensions
      this.pollingInterval = setInterval(this.loadImageDimensions, 20);
    }
  };

  handleImageLoad = () => {
    // Clear dimension polling interval, if for some reason we're still trying to find those
    if (this.pollingInterval) {
      clearInterval(this.pollingInterval);
    }

    // Ensure image dimensions have been loaded
    // Catches case where image is immediately complete, and loadImageDimensions() is not called before this function
    if (!this.state.imageHeight || !this.state.imageWidth) {
      this.loadImageDimensions();
    }

    this.setState({imageLoaded: true});
    this.image = null;
  };

  loadImageDimensions = () => {
    if (this.image && this.image.naturalHeight && this.image.naturalWidth) {
      if (this.pollingInterval) {
        clearInterval(this.pollingInterval);
      }
      this.setState({
        imageWidth: this.image.naturalWidth,
        imageHeight: this.image.naturalHeight,
      });
    }
  };

  renderImage = () => {
    if (this.state.imageLoaded) {
      // NOTE: For the link href, we prioritize the remote url over the local url as Edge will not load local blob URLs.
      // For the image src, we prioritize the local blob URL over the remote URL to avoid unnecessary hits to S3
      return (
        <a
          href={this.props.message.url || this.props.message.localBlobUrl}
          target="_blank"
          rel="noopener noreferrer"
        >
          <img src={this.props.message.localBlobUrl || this.props.message.url} />
        </a>
      );
    }

    if (this.state.imageHeight && this.state.imageWidth) {
      return (
        <canvas
          className="placeholder"
          width={this.state.imageWidth}
          height={this.state.imageHeight}
        />
      );
    }

    return null;
  };

  render() {
    const fromCustomer = this.props.message.authorType === 'Customer';
    const isUploading = this.props.message.status && this.props.message.status === 'pending';

    return (
      <div className={classnames('ImageMessage', {uploading: isUploading, fromCustomer})}>
        {this.renderImage()}
        {isUploading && (
          <div className="progress">
            <ProgressCircle
              percentage={this.props.message.uploadProgress || 0}
              size={Math.min(60, 0.8 * Math.min(this.state.imageWidth, this.state.imageHeight))}
              progressWidth={4}
              trackWidth={4}
              progressColor={'rgba(255, 255, 255, 0.7)'}
              trackColor="rgba(128, 128, 128, 0.6)"
            />
          </div>
        )}
      </div>
    );
  }
}

export default ImageMessage;
