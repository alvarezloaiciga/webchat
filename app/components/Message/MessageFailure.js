import React from 'react';
import {getMessage} from 'reducers/chat';
import {intlMessageTypes} from 'Common/Constants';
import styled, {css} from 'react-emotion';
import Icon from 'core-ui/components/Icon';
import {timesCircle} from 'Icons';
import {connect} from 'react-redux';
import {getStyle} from 'Common/QuiqOptions';
import type {ChatState, ChatConfiguration} from 'Common/types';

export type MessageFailureProps = {
  reason: ?string,
  configuration: ChatConfiguration,
};

const MessageFailureContainer = styled.div`
  width: 100%;
  color: red;
`;

const errorIcon = css`
  margin-right: 3px;
`;

const getReasonMessage = (reason: ?string): ?string => {
  switch (reason) {
    case 'CONTENT_TYPE_NOT_ALLOWED':
      return getMessage(intlMessageTypes.unsupportedFileType);

    default:
      return undefined;
  }
};

export const MessageFailure = (props: MessageFailureProps) => {
  const reasonMessage = getReasonMessage(props.reason);
  const {colors, styles} = props.configuration;
  return (
    <MessageFailureContainer
      style={getStyle(styles.MessageErrorText, {color: colors.error})}
      className="MessageFailure"
    >
      <Icon icon={timesCircle} className={errorIcon} />
      {getMessage(intlMessageTypes.unableToSend)}
      {reasonMessage && `: ${reasonMessage}`}
    </MessageFailureContainer>
  );
};

export default connect((state: ChatState) => ({configuration: state.configuration}), null)(
  MessageFailure,
);
