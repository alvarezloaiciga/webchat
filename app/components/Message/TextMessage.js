// @flow

import React from 'react';
import quiqOptions, {getStyle} from 'Common/QuiqOptions';
import classnames from 'classnames';
import Linkify from 'react-linkify';
import Twemoji from 'react-easy-emoji';
import {isSingleEmoji} from 'utils/emojiUtils';
import type {TextMessage as TextMessageType} from 'Common/types';
import './styles/TextMessage.scss';

export type TextMessageProps = {
  message: TextMessageType,
};

/**
 * Convert Emoji characters to <img> tags pointed at twemoji.
 * @param text {string} - String to emojify.
 * @return Array of React elements
 */
const emojify = (text: string): Array<React$Element<*>> => {
  const singleEmoji = isSingleEmoji(text);
  return Twemoji(text, {
    props: {
      style: {
        height: singleEmoji ? '30px' : '1em',
        width: singleEmoji ? '30px' : '1em',
        margin: '0px 0.05em 0px 0.1em',
        verticalAlign: '-0.2em',
      },
    },
  });
};

const {fontFamily, colors, styles, width} = quiqOptions;

const getMessageBubbleStyle = (fromCustomer: boolean) => {
  if (fromCustomer) {
    return getStyle(styles.CustomerMessageBubble, {
      backgroundColor: colors.customerMessageBackground,
      maxWidth: width - 82,
    });
  }

  return getStyle(styles.AgentMessageBubble, {
    backgroundColor: colors.agentMessageBackground,
    maxWidth: width - 50,
  });
};

const getMessageTextStyle = (fromCustomer: boolean) => {
  if (fromCustomer) {
    return getStyle(styles.CustomerMessageText, {
      color: colors.customerMessageText,
      fontFamily,
    });
  }

  return getStyle(styles.AgentMessageText, {
    color: colors.agentMessageText,
    fontFamily,
  });
};

export const TextMessage = (props: TextMessageProps) => {
  const fromCustomer = props.message.authorType === 'Customer';
  const textStyle = getMessageTextStyle(fromCustomer);

  return (
    <div
      style={getMessageBubbleStyle(fromCustomer)}
      className={classnames('TextMessage', {fromCustomer})}
    >
      <Linkify
        properties={{
          target: '_blank',
          rel: 'noopener noreferrer',
          style: {
            fontFamily,
            textDecoration: 'underline',
            ...textStyle,
            color: fromCustomer ? colors.customerMessageLinkText : colors.agentMessageLinkText,
          },
        }}
      >
        <span style={textStyle}>{emojify(props.message.text)}</span>
      </Linkify>
    </div>
  );
};

export default TextMessage;
