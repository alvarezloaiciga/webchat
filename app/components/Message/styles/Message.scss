.messageContainer {
  margin: 8px 50px 8px 16px;
  display: flex;
  flex: 0 0 auto;
  align-self: stretch;
  overflow-x: hidden;
  overflow-y: hidden;

  &.fromCustomer {
    margin-right: 16px;
    margin-left: 50px;
  }
}

.fromCustomer {
  align-self: stretch;
}

.noFlexbox .messageContainer {
  display: block;
  width: auto;
  min-width: 20px;
  clear: both;
  float: left;

  &.fromCustomer {
    float: right;
  }
}
