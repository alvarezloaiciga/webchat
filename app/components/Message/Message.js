// @flow

import React from 'react';
import {connect} from 'react-redux';
import TextMessage from './TextMessage';
import ImageMessage from './ImageMessage';
import FileMessage from './FileMessage';
import AgentTypingMessage from './AgentTypingMessage';
import classnames from 'classnames';
import {getConfiguration} from 'reducers/chat';
import Avatar from '../Avatar';
import type {Message as MessageType, ChatState, ChatConfiguration} from 'Common/types';
import './styles/Message.scss';

export type MessageProps = {
  message: MessageType,
  scrollToBottom: () => void,
  configuration: ChatConfiguration,
};

export const Message = (props: MessageProps) => {
  const fromCustomer = props.message.authorType === 'Customer';

  const avatarUrl =
    props.message.authorProfilePicture ||
    (fromCustomer
      ? props.configuration.defaultCustomerAvatar
      : props.configuration.defaultAgentAvatar);

  const showAvatar = props.message.type !== 'AgentTyping';

  let messageComponent;
  switch (props.message.type) {
    case 'Text':
      // $FlowIssue - Flow's type refinement isn't working here
      messageComponent = <TextMessage {...props} />;
      break;
    case 'Attachment':
      if (props.message.contentType.startsWith('image/')) {
        // $FlowIssue - Flow's type refinement isn't working here
        messageComponent = <ImageMessage scrollToBottom={props.scrollToBottom} {...props} />;
      } else {
        // $FlowIssue - Flow's type refinement isn't working here
        messageComponent = <FileMessage {...props} />;
      }
      break;
    case 'AgentTyping':
      messageComponent = <AgentTypingMessage {...props} />;
      break;
  }

  return (
    <div className={classnames('messageContainer', {fromCustomer})}>
      {!fromCustomer &&
        showAvatar && (
          <Avatar url={avatarUrl} authorDisplayName={props.message.authorDisplayName} />
        )}
      {messageComponent}
      {fromCustomer &&
        showAvatar && (
          <Avatar url={avatarUrl} authorDisplayName={props.message.authorDisplayName} forCustomer />
        )}
    </div>
  );
};

export default connect(
  (state: ChatState) => ({
    configuration: getConfiguration(state),
  }),
  {},
)(Message);
