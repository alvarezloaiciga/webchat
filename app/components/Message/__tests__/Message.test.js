// @flow

jest.mock('Common/QuiqOptions');

import React from 'react';
import {getMockMessage} from 'utils/testHelpers';
import type {MessageProps} from '../Message';
import Message from '../Message';
import {shallow} from 'enzyme';
import type {ShallowWrapper} from 'enzyme';

describe('Message component', () => {
  let wrapper: ShallowWrapper;
  let testProps: MessageProps;
  let render: () => void;

  beforeEach(() => {
    testProps = {
      message: getMockMessage(),
      scrollToBottom: jest.fn(),
    };
    render = () => {
      wrapper = shallow(<Message {...testProps} />);
    };
  });

  describe('rendering', () => {
    beforeEach(() => {
      render();
    });

    it('renders with default props', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('fromAgent', () => {
    it('colors the message according to quiqOptions', () => {
      testProps.message = getMockMessage(0, {authorType: 'User'});
      render();
      expect(wrapper).toMatchSnapshot();
    });
  });
});
