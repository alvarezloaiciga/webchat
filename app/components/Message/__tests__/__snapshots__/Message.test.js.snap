// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component fromAgent colors the message according to quiqOptions 1`] = `
<div
  className="messageContainer"
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-start",
        "marginRight": "auto",
      }
    }
  >
    <div
      className="agentAvatar"
      style={Object {}}
    />
    <TextMessage
      message={
        Object {
          "authorType": "User",
          "id": "id--0",
          "text": "text--0",
          "timestamp": 0,
          "type": "Text",
        }
      }
      scrollToBottom={[Function]}
    />
  </div>
</div>
`;

exports[`Message component rendering renders with default props 1`] = `
<div
  className="messageContainer fromCustomer"
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "flex-end",
        "marginLeft": "auto",
      }
    }
  >
    <TextMessage
      message={
        Object {
          "authorType": "Customer",
          "id": "id--0",
          "text": "text--0",
          "timestamp": 0,
          "type": "Text",
        }
      }
      scrollToBottom={[Function]}
    />
    <div
      className="customerAvatar"
      style={Object {}}
    />
  </div>
</div>
`;
